### PR TITLE
feat(nuxt): cache vue/nitro build outputs

### LINKED ISSUE #23392: Optimize build time speed with a cache (or other settings)
### Describe the feature

Presently Nuxt seems to lack a build cache? (_[Next has one](https://nextjs.org/docs/pages/building-your-application/deploying/ci-build-caching) that can be cached for speeding up builds_)

In a real Nuxt project, I am observering `nuxt build` taking roughly 38 seconds (_8s vite client, 11s vite server, 19s nitro server_). This does not improve when running `nuxt build` again, even though nothing has changed.

Is it possible to optimize any of the build steps involved?
- Do the build tools used support re-use of a cache and Nuxt just needs to expose a setting (_or document one_) for enabling this support?
- If not, are there any opportunities to implement this where it'd be worthwhile?

Perhaps you could use [BLAKE3](https://github.com/BLAKE3-team/BLAKE3#readme) via [`@noble/hashes`](https://www.npmjs.com/package/@noble/hashes#blake3) as a speedy way to get a checksum on build inputs? (_could be an opt-in flag_). This is perhaps a naive approach, and depending on overhead may not always help reduce the build time 😅 

---

### Reproduction / Comparison

Nuxt example:

```console
# Reproduction environment:
$ docker run --rm -it --workdir /reproduction node:20-bookworm-slim bash
$ corepack enable

# An example project (possibly too small for measurable difference):
$ npx -y nuxi init --packageManager pnpm --gitInit false .
# Repeat this step, no improvement to build time:
$ time pnpm build
```

Next.js example:

```console
# Reproduction environment:
$ docker run --rm -it --workdir /reproduction node:20-bookworm-slim bash
$ corepack enable

# An example project:
$ npx -y create-next-app@latest --use-pnpm --example hello-world .
# Repeat this step, 1st build 10s while builds after are 5s:
$ time pnpm build
```

---

### Related

I've looked through Nuxt and Nitro docs, nothing really stands out cache wise for a CI to optimize reducing future builds with. Ideally a small change (_or none at all_) should not require the full build time.

> Check existing [discussions](https://github.com/nuxt/nuxt/discussions) and [issues](https://github.com/nuxt/nuxt/issues).

First five pages of each for keywords searches of: `build time`, `build speed`, `build cache`:

- Disabling GZIP in builds for some gain (_`sourcemaps` is another potential option_): https://github.com/nuxt/nuxt/issues/23352#issuecomment-1732026846
- 4 minute build times: https://github.com/nuxt/nuxt/discussions/20642
- 6 minute build times: https://github.com/nuxt/nuxt/discussions/20494
- 10 minute report, with 1 minute if disabling SSR: https://github.com/nuxt/nuxt/issues/20042
- 80s to 5-10s report if using `experimental.treeshakeClientOnly: false`: https://github.com/nuxt/nuxt/issues/20042#issuecomment-1495180211
  - https://nuxt.com/docs/guide/going-further/experimental-features#treeshakeclientonly
  - https://nuxt.com/docs/api/configuration/nuxt-config#treeshakeclientonly
- A report of Nuxt spending half build time in nitro, but nitro direct being fast: https://github.com/nuxt/nuxt/issues/20042#issuecomment-1667765867
- A suggestion that upcoming Rollup 4.0 release adopting the rust-based SWC may bring improvements: https://github.com/nuxt/nuxt/issues/20042#issuecomment-1703711265
  - Associated Rollup 4 PR: https://github.com/rollup/rollup/pull/5073
  - Rollup 4 release tracking PR: https://github.com/rollup/rollup/pull/5140
- Mar 2019 **Nuxt 2** doc links for several build config options to optimize build time (_all settings tied to webpack_): https://github.com/nuxt/nuxt/issues/5131#issuecomment-468231314

There is a deployment guide, but not sure if Nuxt has one for minimizing time of production builds.

### Additional information

- [ ] Would you be willing to help implement this feature?
- [ ] Could this feature be implemented as a module?

### Final checks

- [X] Read the [contribution guide](https://nuxt.com/docs/community/contribution).
- [X] Check existing [discussions](https://github.com/nuxt/nuxt/discussions) and [issues](https://github.com/nuxt/nuxt/issues).

### PR DESCRIPTION
**Pull request by danielroe**
### 🔗 Linked issue

resolves https://github.com/nuxt/nuxt/issues/23392
resolves https://github.com/nuxt/nuxt/pull/28594

### 📚 Description

Following up on https://github.com/nuxt/nuxt/pull/28594, this PR implements Nuxt build caching, both for the Vue app (client/server) + the nitro app. (For example, changing Nitro configuration will not cause a rebuild of the Vue app.) It also updates the production `buildDir` to be `node_modules/.cache/nuxt/.nuxt`.

For now, it's highly experimental and should be tested with care.

Build platforms that cache `node_modules` should be immediately compatible, but this needs further investigation to confirm.

> Inspired by - and using significant chunks of the implementation of - https://github.com/pi0/nuxt-build-cache

### Usage

```ts twoslash [nuxt.config.ts]
export default defineNuxtConfig({
  experimental: {
    buildCache: true
  }
})
```

When enabled, changes to the following files will trigger a full rebuild:

```
.nuxtrc
.npmrc
package.json
package-lock.json
yarn.lock
pnpm-lock.yaml
tsconfig.json
bun.lockb
```

In addition, any changes to files within `srcDir` will trigger a rebuild of the Vue client/server bundle, and any files changed within `serverDir` (as well as any changes to the Vue client/server bundle) will trigger a rebuild of the Nitro server.

### 🚧 TODO:
- [x] write build manifest and _clean up_ caches in FIFO pattern: https://github.com/nuxt/nuxt/issues/14605
- [x] avoid rebuild when running `nuxi generate`
- [x] read/write `.tar` files instead

### CODE REVIEW

File: packages/nuxt/src/core/nitro.ts [hunk 1]
```diff
@@ -517,24 +520,54 @@ export async function initNitro (nuxt: Nuxt & { _nitro?: Nitro }) {
...
+    await build(nitro)
+    logger.wrapAll()
+
+    await symlinkDist()
+
+    if (nuxt.options.experimental.buildCache) {
+      const { hash: cacheId } = await getNitroHash(nuxt)
+      const cacheDir = join(nuxt.options.rootDir, 'node_modules/.cache/nuxt/builds/nitro', cacheId)
+      for (const [key, folder] of Object.entries(foldersToCache)) {
+        await cp(folder, join(cacheDir, key), { recursive: true })
```
Comment (danielroe) [inline packages/nuxt/src/core/nitro.ts:570]: @pi0 do you think we need to/should use `nanotar` rather than copying the folders?

Comment (pi0) [reply packages/nuxt/src/core/nitro.ts]: random read/write on filesystem is costly this is for perf (tar is literally a buffer of files so unless random access to cache files is needed i think this is faster)

File: packages/nuxt/src/core/cache.ts [hunk 2]
```diff
@@ -0,0 +1,332 @@
...
+      const elapsed = Date.now() - start
+      if (res) {
+        consola.success(`Restored Vue client and server builds from cache in \`${elapsed}ms\`.`)
+      }
+      return res
+    },
+  }
+}
+
+export async function getNitroHash (nuxt: Nuxt) {
```
Comment (pi0) [inline packages/nuxt/src/core/cache.ts:64]: Nitro will have its internal logic for build-cache collection so I would suggest avoiding caching nitro artifacts (at least in first step, until we find a common language, ideally an API for nitro to announce it's catchable artifacts and their hashes, besides nitro part of the build isn't slowest part today)

**Review by pi0 — APPROVED** [review]
🚀

**Pull request merged**